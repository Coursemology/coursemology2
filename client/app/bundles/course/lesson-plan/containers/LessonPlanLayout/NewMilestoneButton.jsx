import { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  injectIntl,
  defineMessages,
  intlShape,
  FormattedMessage,
} from 'react-intl';
import { Button } from '@mui/material';
import { showMilestoneForm, createMilestone } from 'course/lesson-plan/actions';

const translations = defineMessages({
  newMilestone: {
    id: 'course.lessonPlan.NewMilestoneButton.newMilestone',
    defaultMessage: 'New Milestone',
  },
  success: {
    id: 'course.lessonPlan.NewMilestoneButton.success',
    defaultMessage: 'Milestone created.',
  },
  failure: {
    id: 'course.lessonPlan.NewMilestoneButton.failure',
    defaultMessage: 'Failed to create milestone.',
  },
});

const styles = {
  button: {
    marginRight: 16,
  },
};

class NewMilestoneButton extends Component {
  createMilestoneHandler = (data) => {
    const { dispatch } = this.props;
    const successMessage = <FormattedMessage {...translations.success} />;
    const failureMessage = <FormattedMessage {...translations.failure} />;
    return dispatch(createMilestone(data, successMessage, failureMessage));
  };

  showForm = () => {
    const { dispatch, intl } = this.props;
    return dispatch(
      showMilestoneForm({
        onSubmit: this.createMilestoneHandler,
        formTitle: intl.formatMessage(translations.newMilestone),
        initialValues: {},
      }),
    );
  };

  render() {
    if (!this.props.canManageLessonPlan) {
      return <div />;
    }

    return (
      <Button
        variant="contained"
        color="primary"
        onClick={this.showForm}
        style={styles.button}
      >
        <FormattedMessage {...translations.newMilestone} />
      </Button>
    );
  }
}

NewMilestoneButton.propTypes = {
  canManageLessonPlan: PropTypes.bool.isRequired,

  dispatch: PropTypes.func.isRequired,
  intl: intlShape.isRequired,
};

export default connect((state) => ({
  canManageLessonPlan: state.flags.canManageLessonPlan,
}))(injectIntl(NewMilestoneButton));
