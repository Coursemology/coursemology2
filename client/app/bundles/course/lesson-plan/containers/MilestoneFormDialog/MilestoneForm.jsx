import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { reduxForm, Field, Form } from 'redux-form';
import renderTextField from 'lib/components/redux-form/TextField';
import RichTextField from 'lib/components/redux-form/RichTextField';
import formTranslations from 'lib/translations/form';
import DateTimePicker from 'lib/components/redux-form/DateTimePicker';
import translations from 'course/lesson-plan/translations';
import { formNames, fields } from 'course/lesson-plan/constants';

const { TITLE, DESCRIPTION, START_AT } = fields;

const validate = (values) => {
  const errors = {};

  const requiredFields = ['title', 'start_at'];
  requiredFields.forEach((field) => {
    if (!values[field]) {
      errors[field] = formTranslations.required;
    }
  });

  return errors;
};

const MilestoneForm = ({ handleSubmit, onSubmit, disabled }) => (
  <Form onSubmit={handleSubmit(onSubmit)}>
    <Field
      fullWidth
      name="title"
      label={<FormattedMessage {...translations[TITLE]} />}
      component={renderTextField}
      {...{ disabled }}
    />
    <br />
    <Field
      fullWidth
      name="description"
      label={<FormattedMessage {...translations[DESCRIPTION]} />}
      component={RichTextField}
      multiline
      rows={2}
      {...{ disabled }}
    />
    <Field
      name="start_at"
      label={<FormattedMessage {...translations[START_AT]} />}
      component={DateTimePicker}
      {...{ disabled }}
    />
  </Form>
);

MilestoneForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
};

export default reduxForm({
  form: formNames.MILESTONE,
  validate,
})(MilestoneForm);
