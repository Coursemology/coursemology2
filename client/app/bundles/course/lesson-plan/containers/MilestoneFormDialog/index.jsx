import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { isPristine } from 'redux-form';
import FormDialogue from 'lib/components/FormDialogue';
import { formNames } from 'course/lesson-plan/constants';
import * as actionCreators from 'course/lesson-plan/actions';
import MilestoneForm from './MilestoneForm';

const MilestoneFormDialog = ({
  visible,
  disabled,
  formTitle,
  initialValues,
  onSubmit,
  pristine,
  dispatch,
}) => {
  const { hideMilestoneForm, submitMilestoneForm } = bindActionCreators(
    actionCreators,
    dispatch
  );

  return (
    <FormDialogue
      title={formTitle}
      open={visible}
      submitForm={submitMilestoneForm}
      skipConfirmation={pristine}
      disabled={disabled}
      hideForm={hideMilestoneForm}
    >
      <MilestoneForm {...{ initialValues, onSubmit, disabled }} />
    </FormDialogue>
  );
};

MilestoneFormDialog.defaultProps = {
  visible: false,
  disabled: false,
};

MilestoneFormDialog.propTypes = {
  visible: PropTypes.bool,
  disabled: PropTypes.bool,
  formTitle: PropTypes.string,
  initialValues: PropTypes.shape({
    title: PropTypes.string,
    description: PropTypes.string,
    start_at: PropTypes.string,
  }),
  onSubmit: PropTypes.func.isRequired,
  pristine: PropTypes.bool.isRequired,
  dispatch: PropTypes.func.isRequired,
};

export default connect(({ milestoneForm, ...state }) => ({
  ...milestoneForm,
  pristine: isPristine(formNames.MILESTONE)(state),
}))(MilestoneFormDialog);
