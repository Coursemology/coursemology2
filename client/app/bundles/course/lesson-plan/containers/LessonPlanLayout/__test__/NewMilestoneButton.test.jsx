import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import ReactTestUtils from 'react-dom/test-utils';
import CourseAPI from 'api/course';
import storeCreator from 'course/lesson-plan/store';
import MilestoneFormDialog from 'course/lesson-plan/containers/MilestoneFormDialog';
import NewMilestoneButton from '../NewMilestoneButton';

beforeEach(() => {
  // add window.matchMedia
  // this is necessary for the date picker to be rendered in desktop mode.
  // if this is not provided, the mobile mode is rendered, which might lead to unexpected behavior
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: (query) => ({
      media: query,
      // this is the media query that @material-ui/pickers uses to determine if a device is a desktop device
      matches: query === '(pointer: fine)',
      onchange: () => {},
      addEventListener: () => {},
      removeEventListener: () => {},
      addListener: () => {},
      removeListener: () => {},
      dispatchEvent: () => false,
    }),
  });
});

afterEach(() => {
  delete window.matchMedia;
});

describe('<NewMilestoneButton />', () => {
  it('allows milestone to be created via MilestoneFormDialog', () => {
    const spyCreate = jest.spyOn(CourseAPI.lessonPlan, 'createMilestone');
    const store = storeCreator({ flags: { canManageLessonPlan: true } });
    const contextOptions = buildContextOptions(store);

    const milestoneFormDialog = mount(<MilestoneFormDialog />, contextOptions);
    const newMilestoneButton = mount(<NewMilestoneButton />, contextOptions);

    // Click 'new milestone' button
    newMilestoneButton.find('button').simulate('click');
    milestoneFormDialog.update();
    expect(
      milestoneFormDialog.find('MilestoneFormDialog').first().props().visible,
    ).toBe(true);

    // Fill milestone form
    const milestoneData = {
      title: 'Ambitious milestone title',
      start_at: new Date('2016-12-31T16:00:00.000Z'),
    };
    const startAt = '01-01-2017';
    const milestoneForm = milestoneFormDialog.find('form');
    const titleInput = milestoneForm.find('input[name="title"]');
    titleInput.simulate('change', { target: { value: milestoneData.title } });
    const startAtDateInput = milestoneForm
      .find('input[name="start_at"]')
      .first();
    startAtDateInput.simulate('change', { target: { value: startAt } });
    startAtDateInput.simulate('blur');

    // Submit milestone form
    const submitButton = milestoneFormDialog
      .find('FormDialogue')
      .first()
      .instance().submitButton;
    ReactTestUtils.Simulate.click(ReactDOM.findDOMNode(submitButton));
    expect(spyCreate).toHaveBeenCalledWith({
      lesson_plan_milestone: milestoneData,
    });
  });

  it('is hidden when canManageLessonPlan is false', () => {
    const store = storeCreator({ flags: { canManageLessonPlan: false } });
    const contextOptions = buildContextOptions(store);
    const newMilestoneButton = mount(<NewMilestoneButton />, contextOptions);
    expect(newMilestoneButton).toMatchSnapshot();
  });
});
