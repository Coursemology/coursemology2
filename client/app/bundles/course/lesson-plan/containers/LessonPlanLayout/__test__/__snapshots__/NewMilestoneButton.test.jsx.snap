// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewMilestoneButton /> is hidden when canManageLessonPlan is false 1`] = `
<Connect(injectIntl(NewMilestoneButton))>
  <injectIntl(NewMilestoneButton)
    canManageLessonPlan={false}
    dispatch={[Function]}
  >
    <NewMilestoneButton
      canManageLessonPlan={false}
      dispatch={[Function]}
      intl={
        {
          "$t": [Function],
          "defaultFormats": {},
          "defaultLocale": "en",
          "defaultRichTextElements": undefined,
          "fallbackOnEmptyString": true,
          "formatDate": [Function],
          "formatDateTimeRange": [Function],
          "formatDateToParts": [Function],
          "formatDisplayName": [Function],
          "formatList": [Function],
          "formatListToParts": [Function],
          "formatMessage": [Function],
          "formatNumber": [Function],
          "formatNumberToParts": [Function],
          "formatPlural": [Function],
          "formatRelativeTime": [Function],
          "formatTime": [Function],
          "formatTimeToParts": [Function],
          "formats": {},
          "formatters": {
            "getDateTimeFormat": [Function],
            "getDisplayNames": [Function],
            "getListFormat": [Function],
            "getMessageFormat": [Function],
            "getNumberFormat": [Function],
            "getPluralRules": [Function],
            "getRelativeTimeFormat": [Function],
          },
          "locale": "en",
          "messages": {},
          "onError": [Function],
          "onWarn": [Function],
          "textComponent": Symbol(react.fragment),
          "timeZone": "Asia/Singapore",
          "wrapRichTextChunksInFragment": undefined,
        }
      }
    >
      <div />
    </NewMilestoneButton>
  </injectIntl(NewMilestoneButton)>
</Connect(injectIntl(NewMilestoneButton))>
`;
